src/typechecker.erl

```diff
@@ -3486,15 +3491,18 @@ refinable(RefinableTy = {user_type, _Anno, _Name, _Args}, TEnv, Trace) ->
             %% Refinability will be determined by the variants which are not (mutually) recursive.
             true;
         false ->
-            case gradualizer_lib:get_user_type_definition(RefinableTy, TEnv#tenv.types) of
+            case gradualizer_lib:get_type_definition(RefinableTy, tenv_to_map(TEnv)) of
                 {ok, Ty} -> refinable(Ty, TEnv, sets:add_element(RefinableTy, Trace));
-                opaque -> false;
+                opaque -> true;
                 not_found -> false
             end
     end;
 refinable(_, _, _) ->
     false.
 
+tenv_to_map(#tenv{module = M, types = Tys, records = Rs}) ->
+    #{module => M, types => Tys, records => Rs}.
```

Comment:
So now we have two representations of TEnv... Not ideal but I can live with it if there is a plan to unify them in the future.  I'd prefer either of

1. Change TEnv to a map everywhere.
2. Use the record everywhere and move it to an include file.

Reply:
I'd prefer changing it to a map everywhere, unless you think it's better to keep it as a record. I can do this in this PR or a separate one.

Reply:
BTW, this map for the time being is described by https://github.com/erszcz/Gradualizer/blob/exhaustive-user-type/src/gradualizer_lib.erl#L87-L90 - where do you think is the right place to put this type?

Reply:
It's internal so any other module than `gradualizer`. I suppose `gradualizer_lib` is fine.

The modules are a bit messy. Some are prefixed by `gradualizer_` and some aren't. The `typechecker` module has `export_all`. We should clean up this mess some day...

Reply:
> I'd prefer changing it to a map everywhere, unless you think it's better to keep it as a record. I can do this in this PR or a separate one.

I guess a separate PR would be better, if it's not too much to ask. It can be done either before or after this PR is merged, as long as it's not forgotten.

Reply:
Done with #333.